src/NuGet.Server/Publishing/PackageService.cs

```diff
@@ -78,19 +80,32 @@ public void DeletePackage(HttpContextBase context)
 
         public void DownloadPackage(HttpContextBase context)
         {
-            RouteData routeData = GetRouteData(context);
+            var routeData = GetRouteData(context);
             // Get the package file name from the route
-            string packageId = routeData.GetRequiredString("packageId");
+            var packageId = routeData.GetRequiredString("packageId");
             var version = new SemanticVersion(routeData.GetRequiredString("version"));
 
-            IPackage requestedPackage = _serverRepository.FindPackage(packageId, version);
-
+            var requestedPackage = _serverRepository.FindPackage(packageId, version);
             if (requestedPackage != null)
             {
-                Package packageMetatada = _serverRepository.GetMetadataPackage(requestedPackage);
-                context.Response.AddHeader("content-disposition", String.Format("attachment; filename={0}", packageMetatada.Path));
-                context.Response.ContentType = "application/zip";
-                context.Response.TransmitFile(packageMetatada.FullPath);
+                context.Response.AddHeader("content-disposition", 
+                    string.Format("attachment; filename={0}.{1}.nupkg", requestedPackage.Id, requestedPackage.Version.ToNormalizedString()));
+                context.Response.ContentType = "binary/octet-stream";
+
+                var serverPackage = requestedPackage as ServerPackage;
+                if (serverPackage != null && !string.IsNullOrEmpty(serverPackage.FullPath))
+                {
+                    // FullPath known - send the file as-is
+                    context.Response.TransmitFile(serverPackage.FullPath);
+                }
+                else
+                {
+                    // FullPath unknown - stream it
+                    using (var packageStream = requestedPackage.GetStream())
```

Comment:
Since the server implementation can be plugged in/extended, people could provide us with a package repo that in the end does not have the FullPath available.